File: python/core/application/Application.py
"""
Python Application Bootstrap
Refactored Python application using flexible architecture with DI
"""

import asyncio
import logging
import sys
from pathlib import Path

# Add project root to path
project_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(project_root))

from core.container.ServiceContainer import ServiceContainer
from core.config.ConfigurationManager import ConfigurationManager
from adapters.AzureOpenAIAdapter import AzureOpenAIAdapter
from services.ModernAgentOrchestrator import ModernAgentOrchestrator
from services.ResearchService import ResearchService
from utils.logger import setup_logging

class PythonApplication:
    """Modern Python application with flexible architecture"""
    
    def __init__(self):
        self.container = ServiceContainer()
        self.config = None
        self.logger = None
        self.orchestrator = None
        self.is_initialized = False
        self.is_running = False

    async def initialize(self):
        """Initialize the application with DI architecture"""
        try:
            # Setup logging first
            self.logger = setup_logging()
            self.logger.info("🐍 Initializing Python Application with modern architecture...")

            # Phase 1: Configuration Management
            await self._initialize_configuration()
            
            # Phase 2: Register Services
            await self._register_services()
            
            # Phase 3: Initialize Core Services
            await self._initialize_services()
            
            self.is_initialized = True
            self.logger.info("✅ Python Application initialized successfully")
            
        except Exception as e:
            if self.logger:
                self.logger.error(f"❌ Failed to initialize Python application: {e}")
            raise

    async def _initialize_configuration(self):
        """Initialize configuration management"""
        self.config = ConfigurationManager()
        await self.config.initialize()
        
        # Register in container
        self.container.instance('ConfigurationManager', self.config)

    async def _register_services(self):
        """Register services in DI container"""
        
        # LLM Provider — create once and register as instance (singleton)
        config = self.container.resolve('ConfigurationManager')
        llm_config = config.get_section('llm')
        provider_type = llm_config.get('provider', 'azure-openai')
        
        if provider_type == 'azure-openai':
            llm_provider = AzureOpenAIAdapter(self.container)
        elif provider_type == 'claude':
            from adapters.ClaudeAdapter import ClaudeAdapter
            llm_provider = ClaudeAdapter(self.container)
        else:
            raise ValueError(f"Unknown LLM provider: {provider_type}")
        
        self.container.instance('ILLMProvider', llm_provider)
        
        # Agent Orchestrator
        self.container.singleton('AgentOrchestrator', ModernAgentOrchestrator)
        
        # Research Service (web research with DuckDuckGo + BeautifulSoup)
        self.container.instance('IResearchService', ResearchService(self.container))
        
        self.logger.info("✅ Services registered in DI container")

    async def _initialize_services(self):
        """Initialize core services"""
        
        # Initialize LLM Provider
        llm_provider = self.container.resolve('ILLMProvider')
        await llm_provider.initialize()
        
        # Initialize Agent Orchestrator
        self.orchestrator = self.container.resolve('AgentOrchestrator')
        await self.orchestrator.initialize()
        
        self.logger.info("✅ Core services initialized")

    async def start(self):
        """Start the application"""
        if not self.is_initialized:
            raise RuntimeError("Application must be initialized before starting")
        
        try:
            self.logger.info("🚀 Starting Python Application...")
            
            # Start message processing loop
            await self._start_message_loop()
            
            self.is_running = True
            self.logger.info("✅ Python Application started successfully")
            
        except Exception as e:
            self.logger.error(f"❌ Failed to start Python application: {e}")
            raise

    async def _start_message_loop(self):
        """Start the message processing loop for Node.js bridge"""
        self.logger.info("🔄 Starting message processing loop...")
        
        while True:
            try:
                # Read from stdin (Node.js bridge communication)
                line = await asyncio.get_event_loop().run_in_executor(
                    None, sys.stdin.readline
                )
                
                if not line:
                    break
                    
                # Process the request
                await self._handle_bridge_request(line.strip())
                
            except KeyboardInterrupt:
                break
            except Exception as e:
                self.logger.error(f"❌ Error in message loop: {e}")

    async def _handle_bridge_request(self, request_line: str):
        """Handle request from Node.js bridge"""
        request = None
        request_id = 'unknown'
        try:
            import json
            
            request = json.loads(request_line)
            request_id = request.get('id', 'unknown')
            request_type = request.get('type')
            data = request.get('data', {})
            
            self.logger.debug(f"📥 Processing request {request_id}: {request_type}")
            
            result = None
            
            if request_type == 'process_message':
                result = await self.orchestrator.process_message(
                    data.get('message', ''),
                    data.get('context', {})
                )
            elif request_type == 'web_research':
                # Dedicated research request from Node.js WebSearchPlugin
                research_service = self.container.resolve('IResearchService')
                result = await research_service.research(
                    data.get('query', ''),
                    data.get('options', {})
                )
            elif request_type == 'health_check':
                result = await self._health_check()
            elif request_type == 'get_stats':
                result = await self._get_stats()
            else:
                raise ValueError(f"Unknown request type: {request_type}")
            
            # Send response
            response = {
                'id': request_id,
                'result': result,
                'status': 'success',
                'timestamp': asyncio.get_event_loop().time()
            }
            
            print(json.dumps(response))
            sys.stdout.flush()
            
        except Exception as e:
            # Send error response
            error_response = {
                'id': request_id,
                'error': str(e),
                'status': 'error',
                'timestamp': asyncio.get_event_loop().time()
            }
            
            print(json.dumps(error_response))
            sys.stdout.flush()
            
            if self.logger:
                self.logger.error(f"❌ Request processing failed: {e}")
            else:
                print(f"[Python] Request processing failed: {e}", file=sys.stderr)

    async def stop(self):
        """Stop the application gracefully"""
        if self.logger:
            self.logger.info("🛑 Stopping Python Application...")
        else:
            print("[Python] Stopping Python Application...", file=sys.stderr)
        self.is_running = False
        if self.logger:
            self.logger.info("✅ Python Application stopped")
        else:
            print("[Python] Python Application stopped", file=sys.stderr)

    async def _health_check(self):
        """Perform health check"""
        try:
            llm_provider = self.container.resolve('ILLMProvider')
            llm_status = await llm_provider.get_health_status()
            
            return {
                'status': 'healthy',
                'llm_provider': llm_status,
                'services': self.container.get_registered_services(),
                'config_summary': self.config.get_summary()
            }
            
        except Exception as e:
            return {
                'status': 'unhealthy',
                'error': str(e)
            }

    async def _get_stats(self):
        """Get application statistics"""
        return {
            'initialized': self.is_initialized,
            'running': self.is_running,
            'services': len(self.container.get_registered_services()),
            'config_summary': self.config.get_summary() if self.config else {},
            'orchestrator_metrics': await self.orchestrator.get_metrics() if self.orchestrator else {}
        }

    def get_container(self):
        """Get the DI container"""
        return self.container

    def get_config(self):
        """Get the configuration manager"""
        return self.config

# Main entry point
async def main():
    """Main application entry point"""
    app = PythonApplication()
    
    try:
        await app.initialize()
        await app.start()
        
    except KeyboardInterrupt:
        print("\n[Python] Received interrupt signal", file=sys.stderr)
    except Exception as e:
        print(f"[Python] Application error: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        await app.stop()

if __name__ == "__main__":
    asyncio.run(main())

File: nodejs/src/core/application/Application.js
/**
 * Modern Application Bootstrap
 * Refactored main application using new flexible architecture
 * Implements all 6 architectural principles with dependency injection
 */

const ServiceContainer = require('../container/ServiceContainer');
const ConfigurationManager = require('../config/ConfigurationManager');
const PluginManager = require('../../plugins/PluginManager');
const APIVersionManager = require('../../api/versioning/APIVersionManager');

// Database Adapters
const SQLiteAdapter = require('../../database/adapters/SQLiteAdapter');
const PostgreSQLAdapter = require('../../database/adapters/PostgreSQLAdapter');

// LLM Adapters
const AzureOpenAIAdapter = require('../../llm/adapters/AzureOpenAIAdapter');
const ClaudeAdapter = require('../../llm/adapters/ClaudeAdapter');

// Message Provider Adapters
const WhatsAppAdapter = require('../../messaging/adapters/WhatsAppAdapter');

// Services (refactored to use DI)
const MessageHandlerService = require('../../services/MessageHandlerService');
const MediaHandlerService = require('../../services/MediaHandlerService');
const AdminService = require('../../services/AdminService');
const MetricsService = require('../../services/MetricsService');
const NotificationService = require('../../services/NotificationService');

// Media Processing Adapters
const AzureMediaAdapter = require('../../media/adapters/AzureMediaAdapter');
const LocalWhisperAdapter = require('../../media/adapters/LocalWhisperAdapter');
const GroqMediaAdapter = require('../../media/adapters/GroqMediaAdapter');

// TTS Adapters
const GeminiTTSAdapter = require('../../tts/adapters/GeminiTTSAdapter');

// Bridge Services
const ModernPythonBridge = require('../../bridge/ModernPythonBridge');

// Dashboard & Auth
const DashboardServer = require('../../dashboard/server');
const AdminAuthService = require('../../services/admin_auth.service');

const logger = require('../../utils/logger');

class Application {
    constructor() {
        this.container = new ServiceContainer();
        this.config = null;
        this.pluginManager = null;
        this.apiVersionManager = null;
        this.isInitialized = false;
        this.isRunning = false;
        this.startTime = null;
        this.gracefulShutdown = this.gracefulShutdown.bind(this);
    }

    async initialize() {
        try {
            logger.info('🚀 Initializing modern WhatsApp AutoGen Bot...');
            this.startTime = Date.now();

            // Phase 1: Configuration Management
            await this.initializeConfiguration();

            // Phase 2: Register Core Services
            await this.registerServices();

            // Phase 3: Plugin System
            await this.initializePlugins();

            // Phase 4: API Versioning
            await this.initializeAPIVersioning();

            // Phase 5: Setup Graceful Shutdown
            this.setupGracefulShutdown();

            this.isInitialized = true;

            const initTime = Date.now() - this.startTime;
            logger.info(`✅ Application initialized in ${initTime}ms`);

        } catch (error) {
            logger.error('❌ Failed to initialize application:', error);
            throw error;
        }
    }

    async start() {
        if (!this.isInitialized) {
            throw new Error('Application must be initialized before starting');
        }

        try {
            logger.info('🎬 Starting WhatsApp AutoGen Bot...');

            // Initialize database connection
            // const database = this.container.resolve('IDatabase');
            // await database.connect();
            // await database.initializeSchema();

            // // Initialize LLM provider
            // const llmProvider = this.container.resolve('ILLMProvider');
            // await llmProvider.initialize();
            // await llmProvider.testConnection();

            // // Initialize message provider
            // const messageProvider = this.container.resolve('IMessageProvider');
            // await messageProvider.initialize();
            // await messageProvider.connect();
            await this.initializeProviders();
            // Start core services
            await this.startServices();

            // Start web dashboard (after providers + services are up)
            await this.startDashboard();

            this.isRunning = true;

            const totalTime = Date.now() - this.startTime;
            logger.info(`🚀 WhatsApp AutoGen Bot started successfully in ${totalTime}ms`);

            // Emit ready event
            this.container.resolve('EventEmitter').emit('application:ready');

        } catch (error) {
            logger.error('❌ Failed to start application:', error);
            throw error;
        }
    }

    async stop() {
        if (!this.isRunning) return;

        try {
            logger.info('🛑 Stopping WhatsApp AutoGen Bot...');

            this.isRunning = false;

            // Stop services in reverse order
            await this.stopServices();

            // Disconnect providers
            const messageProvider = this.container.resolve('IMessageProvider');
            await messageProvider.disconnect();

            const database = this.container.resolve('IDatabase');
            await database.disconnect();

            // Cleanup plugin manager
            if (this.pluginManager) {
                // Plugin manager cleanup would go here
            }

            // Cleanup configuration
            if (this.config) {
                await this.config.cleanup();
            }

            logger.info('✅ Application stopped gracefully');

        } catch (error) {
            logger.error('❌ Error during application shutdown:', error);
            throw error;
        }
    }

    async initializeConfiguration() {
        logger.info('⚙️ Initializing configuration management...');

        this.config = new ConfigurationManager({
            environment: process.env.NODE_ENV || 'development'
        });

        await this.config.initialize();

        // Register config in container
        this.container.instance('ConfigurationManager', this.config);

        logger.info('✅ Configuration management initialized');
    }

    async registerServices() {
        logger.info('🔧 Registering services...');

        // Register EventEmitter
        const EventEmitter = require('events');
        this.container.instance('EventEmitter', new EventEmitter());

        // Database Service Factory
        this.container.singleton('IDatabase', (container) => {
            const config = container.resolve('ConfigurationManager');
            const dbConfig = config.getSection('database');

            switch (dbConfig.type) {
                case 'postgresql':
                    return new PostgreSQLAdapter(dbConfig.postgresql);
                case 'sqlite':
                default:
                    return new SQLiteAdapter(dbConfig.sqlite);
            }
        });

        // LLM Provider Factory
        this.container.singleton('ILLMProvider', (container) => {
            const config = container.resolve('ConfigurationManager');
            const llmConfig = config.getSection('llm');

            switch (llmConfig.provider) {
                case 'claude':
                    return new ClaudeAdapter(llmConfig.claude);
                case 'azure-openai':
                default:
                    return new AzureOpenAIAdapter(llmConfig.azureOpenAI);
            }
        });

        // Message Provider Factory
        this.container.singleton('IMessageProvider', (container) => {
            const config = container.resolve('ConfigurationManager');
            const messagingConfig = config.getSection('messaging');

            switch (messagingConfig.provider) {
                case 'whatsapp':
                default:
                    return new WhatsAppAdapter(messagingConfig.whatsapp);
            }
        });

        // Python Bridge
        this.container.singleton('PythonBridge', ModernPythonBridge);

        // Media Processor Factory
        this.container.singleton('IMediaProcessor', (container) => {
            const config = container.resolve('ConfigurationManager');
            const mediaConfig = config.getSection('media') || {};
            const provider = mediaConfig.provider || process.env.MEDIA_PROVIDER || 'azure';

            switch (provider.toLowerCase()) {
                case 'local':
                case 'whisper':
                    return new LocalWhisperAdapter(mediaConfig.local || {});
                case 'groq':
                    return new GroqMediaAdapter(mediaConfig.groq || {});
                case 'azure':
                default:
                    return new AzureMediaAdapter(mediaConfig.azure || {});
            }
        });

        // TTS Provider
        this.container.singleton('ITTSProvider', (container) => {
            const config = container.resolve('ConfigurationManager');
            const ttsConfig = config.getSection('tts') || {};
            return new GeminiTTSAdapter(ttsConfig.gemini || {});
        });

        // Core Services
        this.container.singleton('MessageHandlerService', MessageHandlerService);
        this.container.singleton('MediaHandlerService', MediaHandlerService);
        this.container.singleton('AdminService', AdminService);
        this.container.singleton('MetricsService', MetricsService);
        this.container.singleton('NotificationService', NotificationService);

        logger.info('✅ Services registered');
    }
    async initializeProviders() {
        logger.info('🔌 Initializing providers...');

        try {
            // Initialize database connection FIRST
            logger.info('📊 Connecting to database...');
            const database = this.container.resolve('IDatabase');
            await database.connect();
            logger.info('✅ Database connected');

            await database.initializeSchema();
            logger.info('✅ Database schema initialized');

            // Initialize LLM provider
            logger.info('🤖 Initializing LLM provider...');
            const llmProvider = this.container.resolve('ILLMProvider');
            await llmProvider.initialize();
            await llmProvider.testConnection();
            logger.info('✅ LLM provider initialized');

            // Initialize message provider
            logger.info('📱 Initializing message provider...');
            const messageProvider = this.container.resolve('IMessageProvider');
            await messageProvider.initialize();
            await messageProvider.connect();
            logger.info('✅ Message provider initialized');

            // Initialize TTS provider
            logger.info('🎤 Initializing TTS provider...');
            const ttsProvider = this.container.resolve('ITTSProvider');
            await ttsProvider.initialize();
            logger.info('✅ TTS provider initialized');

            logger.info('✅ All providers initialized and connected');

        } catch (error) {
            logger.error('❌ Failed to initialize providers:', error);
            throw error;
        }
    }
    async initializePlugins() {
        const pluginConfig = this.config.getSection('plugins');

        if (!pluginConfig.enabled) {
            logger.info('🔌 Plugin system disabled');
            return;
        }

        logger.info('🔌 Initializing plugin system...');

        this.pluginManager = new PluginManager(this.container);
        await this.pluginManager.initialize();

        // Register in container
        this.container.instance('PluginManager', this.pluginManager);

        logger.info(`✅ Plugin system initialized with ${this.pluginManager.getStatistics().totalPlugins} plugins`);
    }

    async initializeAPIVersioning() {
        const apiConfig = this.config.getSection('api');

        if (!apiConfig.versioning.enabled) {
            logger.info('📝 API versioning disabled');
            return;
        }

        logger.info('📝 Initializing API versioning...');

        this.apiVersionManager = new APIVersionManager();

        // Register supported versions
        for (const version of apiConfig.versioning.supportedVersions) {
            this.apiVersionManager.registerVersion(version, {
                released: new Date(),
                changes: [`Initial ${version} release`]
            });
        }

        // Register in container
        this.container.instance('APIVersionManager', this.apiVersionManager);

        logger.info('✅ API versioning initialized');
    }

    async startDashboard() {
        const dashboardEnabled = process.env.DASHBOARD_ENABLED !== 'false';
        if (!dashboardEnabled) {
            logger.info('🌐 Web dashboard disabled');
            return;
        }

        try {
            logger.info('🌐 Starting web dashboard...');

            const database = this.container.resolve('IDatabase');
            const messageProvider = this.container.resolve('IMessageProvider');

            // Create auth service with the database
            const authService = new AdminAuthService(database);
            this.container.instance('AdminAuthService', authService);

            // Create dashboard with WhatsApp client, database, and auth
            const whatsappClient = messageProvider.client;
            this.dashboard = new DashboardServer(whatsappClient, database, authService);
            this.dashboard.container = this.container;

            const dashboardPort = parseInt(process.env.DASHBOARD_PORT) || 3000;
            await this.dashboard.start(dashboardPort);

            this.container.instance('DashboardServer', this.dashboard);
            logger.info(`✅ Web dashboard running on http://localhost:${dashboardPort}`);

        } catch (error) {
            logger.warn('⚠️ Failed to start web dashboard (non-fatal):', error.message);
        }
    }

    async startServices() {
        logger.info('🎯 Starting core services...');

        const services = [
            'MetricsService',
            'NotificationService',
            'AdminService',
            'PythonBridge',  // Start Python bridge first
            'MediaHandlerService',
            'MessageHandlerService'
        ];

        // Non-critical services that should not crash the app if they fail
        const optionalServices = ['PythonBridge', 'MediaHandlerService'];

        for (const serviceName of services) {
            try {
                const service = this.container.resolve(serviceName);

                if (service.start && typeof service.start === 'function') {
                    await service.start();
                }

                logger.debug(`✅ ${serviceName} started`);
            } catch (error) {
                if (optionalServices.includes(serviceName)) {
                    logger.warn(`⚠️ Optional service ${serviceName} failed to start (will use fallback):`, error.message);
                } else {
                    logger.error(`❌ Failed to start ${serviceName}:`, error);
                    throw error;
                }
            }
        }

        logger.info('✅ Core services started');
    }

    async stopServices() {
        logger.info('🛑 Stopping core services...');

        const services = [
            'MessageHandlerService',
            'PythonBridge',
            'AdminService',
            'NotificationService',
            'MetricsService'
        ];

        for (const serviceName of services) {
            try {
                const service = this.container.resolve(serviceName);

                if (service.stop && typeof service.stop === 'function') {
                    await service.stop();
                }

                logger.debug(`✅ ${serviceName} stopped`);
            } catch (error) {
                logger.warn(`⚠️ Error stopping ${serviceName}:`, error);
            }
        }

        logger.info('✅ Core services stopped');
    }

    setupGracefulShutdown() {
        // Handle process signals
        process.on('SIGTERM', this.gracefulShutdown);
        process.on('SIGINT', this.gracefulShutdown);
        process.on('SIGUSR1', this.gracefulShutdown);
        process.on('SIGUSR2', this.gracefulShutdown);

        // Handle uncaught exceptions
        process.on('uncaughtException', (error) => {
            logger.error('💥 Uncaught Exception:', error);
            this.gracefulShutdown(1);
        });

        process.on('unhandledRejection', (reason, promise) => {
            logger.error('💥 Unhandled Rejection at:', promise, 'reason:', reason);
            this.gracefulShutdown(1);
        });

        logger.debug('🛡️ Graceful shutdown handlers registered');
    }

    async gracefulShutdown(exitCode = 0) {
        logger.info('🛑 Graceful shutdown initiated...');

        try {
            await this.stop();
            process.exit(exitCode);
        } catch (error) {
            logger.error('❌ Error during graceful shutdown:', error);
            process.exit(1);
        }
    }

    // Utility methods
    getContainer() {
        return this.container;
    }

    getConfig() {
        return this.config;
    }

    getPluginManager() {
        return this.pluginManager;
    }

    getAPIVersionManager() {
        return this.apiVersionManager;
    }

    isHealthy() {
        return this.isInitialized && this.isRunning;
    }

    getStatus() {
        return {
            initialized: this.isInitialized,
            running: this.isRunning,
            uptime: this.startTime ? Date.now() - this.startTime : 0,
            services: this.container.getRegisteredServices(),
            plugins: this.pluginManager ? this.pluginManager.getStatistics() : null,
            config: this.config ? this.config.getSummary() : null
        };
    }

    // Health check endpoint
    async healthCheck() {
        const status = {
            status: 'ok',
            timestamp: new Date().toISOString(),
            uptime: Date.now() - this.startTime,
            services: {},
            providers: {}
        };

        try {
            // Check database
            const database = this.container.resolve('IDatabase');
            status.providers.database = await database.isConnected() ? 'healthy' : 'unhealthy';

            // Check LLM provider
            const llmProvider = this.container.resolve('ILLMProvider');
            status.providers.llm = await llmProvider.isReady() ? 'healthy' : 'unhealthy';

            // Check message provider
            const messageProvider = this.container.resolve('IMessageProvider');
            status.providers.messaging = await messageProvider.isReady() ? 'healthy' : 'unhealthy';

            // Overall health
            const allHealthy = Object.values(status.providers).every(s => s === 'healthy');
            status.status = allHealthy ? 'ok' : 'degraded';

        } catch (error) {
            status.status = 'error';
            status.error = error.message;
        }

        return status;
    }

    // Configuration hot-reload
    async reloadConfiguration() {
        logger.info('🔄 Reloading configuration...');

        try {
            await this.config.reload();

            // Notify services of config change
            const eventEmitter = this.container.resolve('EventEmitter');
            eventEmitter.emit('config:reloaded');

            logger.info('✅ Configuration reloaded');
        } catch (error) {
            logger.error('❌ Failed to reload configuration:', error);
            throw error;
        }
    }

    // Plugin hot-reload
    async reloadPlugins() {
        if (!this.pluginManager) {
            throw new Error('Plugin system not initialized');
        }

        logger.info('🔌 Reloading plugins...');

        try {
            await this.pluginManager.autoLoadPlugins();

            const eventEmitter = this.container.resolve('EventEmitter');
            eventEmitter.emit('plugins:reloaded');

            logger.info('✅ Plugins reloaded');
        } catch (error) {
            logger.error('❌ Failed to reload plugins:', error);
            throw error;
        }
    }
}

module.exports = Application;
